In [4]:
import pandas as pd
import numpy as np

# Load the dataset
# skiprows=15: skips the initial descriptive text
# nrows=744: reads the exact months from July 1963 to June 2025
df = pd.read_csv('25_Portfolios_5x5.csv', skiprows=15, nrows=744)

# Rename the first column to 'Date'
df.rename(columns={df.columns[0]: 'Date'}, inplace=True)

# Convert the Date column to integer format
df['Date'] = df['Date'].astype(int)

# Divide by 100 to convert percentage returns to decimals (e.g., 0.01) 
df.iloc[:, 1:] = df.iloc[:, 1:].astype(float) / 100

# Set Date as the index
df.set_index('Date', inplace=True)

print("Data loaded successfully.")
df.head()

Data loaded successfully.


SMALL LoBM   ME1 BM2   ME1 BM3   ME1 BM4  SMALL HiBM   ME2 BM1  \
Date                                                                     
192607    0.058276 -0.017006  0.005118 -0.021477    0.019583  0.012118   
192608   -0.020206 -0.080282  0.013968  0.021483    0.085104  0.023620   
192609   -0.048291 -0.026806 -0.043417 -0.032683    0.008586 -0.026849   
192610   -0.093633 -0.035519 -0.035024  0.034413   -0.025452 -0.028014   
192611    0.055888  0.041877  0.024384 -0.044495    0.005110  0.031023   

         ME2 BM2   ME2 BM3   ME2 BM4   ME2 BM5  ...   ME4 BM1   ME4 BM2  \
Date                                            ...                       
192607  0.024107  0.006056 -0.026082 -0.004527  ...  0.015376  0.015460   
192608 -0.007525  0.038984  0.002299  0.062937  ...  0.013858  0.038587   
192609 -0.005252  0.010789 -0.032877 -0.009419  ...  0.016897 -0.005246   
192610 -0.044191 -0.050767 -0.080271 -0.013213  ... -0.039136 -0.026528   
192611 -0.017317  0.030425  0.049538  0.027292  ...  0.034492  0.023823   

         ME4 BM3   ME4 BM4   ME4 BM5  BIG LoBM   ME5 BM2   ME5 BM3   ME5 BM4  \
Date                                                                           
192607  0.013389  0.002765  0.024678  0.033248  0.060909  0.020285  0.031263   
192608  0.019738  0.021336  0.053422  0.010169  0.041975  0.019769  0.054924   
192609 -0.017724  0.014806  0.008730 -0.012951  0.036610  0.001384 -0.007497   
192610 -0.021058 -0.032532 -0.053525 -0.027382 -0.030061 -0.022467 -0.046725   
192611  0.037315  0.051102  0.018213  0.044331  0.025355  0.015280  0.036596   

        BIG HiBM  
Date              
192607  0.005623  
192608  0.077576  
192609 -0.024284  
192610 -0.058129  
192611  0.025636  

[5 rows x 25 columns]

### Legend of the 25 Portfolios (Fama-French 5x5)

The test assets used are the 25 Fama-French portfolios formed on Size and Book-to-Market (B/M) ratio. The analysis covers the period from July 1963 to June 2025.

#### Size (Market Equity - ME): Table Rows
The companies are divided into 5 quintiles based on market capitalization:
* **ME1 (Small)**: The smallest companies.
* **ME2, ME3, ME4**: Mid-sized companies.
* **ME5 (Big)**: The largest companies (Large-Cap).

#### Value (Book-to-Market - B/M): Table Columns
The companies are divided into 5 quintiles based on the ratio between book value and market value:
* **BM1 (Low / Growth)**: Companies with high prices relative to book value (e.g., tech sectors).
* **BM2, BM3, BM4**: Intermediate valuation levels.
* **BM5 (High / Value)**: Companies with low prices relative to book value (Value stocks).

**Calculation Note**: All final results are rounded to 2 decimal places as required by the assignment specifications.

In [6]:
# Filter for the specific period: July 1963 - June 2025
# Note: Ensure the index is integer-based (e.g., 196307)
df_filtered = df.loc[196307:202506]

# Calculate arithmetic average monthly returns for each portfolio
mean_returns = df_filtered.mean()

# Reshape the 25 portfolios into a 5x5 table
# Rows = Size quintiles, Columns = B/M quintiles
table_1a = pd.DataFrame(
    mean_returns.values.reshape(5, 5),
    index=['Small', 'Size 2', 'Size 3', 'Size 4', 'Big'],
    columns=['Low B/M', 'B/M 2', 'B/M 3', 'B/M 4', 'High B/M']
)

print("Table 1a: Arithmetic Average Monthly Returns (1963:07 - 2025:06)")
table_1a.round(4)

Table 1a: Arithmetic Average Monthly Returns (1963:07 - 2025:06)


Low B/M   B/M 2   B/M 3   B/M 4  High B/M
Small    0.0092  0.0129  0.0137  0.0150    0.0172
Size 2   0.0093  0.0127  0.0142  0.0152    0.0168
Size 3   0.0093  0.0121  0.0130  0.0141    0.0155
Size 4   0.0092  0.0092  0.0121  0.0136    0.0146
Big      0.0076  0.0084  0.0088  0.0105    0.0111

### Question 1a: Arithmetic Average Returns

The table below reports the average monthly returns for the 25 portfolios formed on Size and Book-to-Market.
The sample period covers **July 1963 to June 2025**.

#### Legend for the 25 Portfolios

**Rows: Size (Market Equity)**
The rows represent 5 groups based on the company's market capitalization:
* **Small**: The smallest 20% of firms (Small-Cap).
* **Size 2, 3, 4**: Mid-sized firms.
* **Big**: The largest 20% of firms (Large-Cap / Blue Chips).

**Columns: Book-to-Market (B/M)**
The columns represent 5 groups based on the ratio of Book Value to Market Value:
* **Low B/M**: "Growth" stocks. These firms have high market prices relative to their book value (e.g., tech companies).
* **B/M 2, 3, 4**: Intermediate values.
* **High B/M**: "Value" stocks. These firms have low market prices relative to their book value (often considered undervalued).

**Note:** Returns are expressed as decimals (e.g., 0.0172 = 1.72%).

Key observations to look for:
* **Size Effect**: Do Small firms outperform Big firms?
* **Value Effect**: Do High B/M (Value) firms outperform Low B/M (Growth) firms?

In [7]:
# --- Automated Analysis Script ---

# 1. Analyze Size Effect (Small vs Big)
# We compare the average return of the Small row (index 0) vs Big row (index 4)
small_avg = table_1a.iloc[0].mean()
big_avg = table_1a.iloc[4].mean()
size_premium = small_avg - big_avg

# 2. Analyze Value Effect (Value vs Growth)
# We compare the average return of the High B/M column (index 4) vs Low B/M column (index 0)
value_avg = table_1a.iloc[:, 4].mean()
growth_avg = table_1a.iloc[:, 0].mean()
value_premium = value_avg - growth_avg

print("### AUTOMATED ANSWER GENERATOR ###\n")

# Generating Answer for Size Effect
print("QUESTION 1: Do Small firms outperform Big firms?")
if size_premium > 0:
    print(f"YES. CONFIRMED.")
    print(f"Data shows that Small firms (Avg: {small_avg:.4f}) have higher returns than Big firms (Avg: {big_avg:.4f}).")
    print(f"The 'Size Premium' is approximately {size_premium*100:.2f}% per month.\n")
else:
    print(f"NO. In this period, Big firms outperformed Small firms.\n")

# Generating Answer for Value Effect
print("QUESTION 2: Do High B/M (Value) firms outperform Low B/M (Growth) firms?")
if value_premium > 0:
    print(f"YES. CONFIRMED.")
    print(f"Data shows that Value firms (Avg: {value_avg:.4f}) have higher returns than Growth firms (Avg: {growth_avg:.4f}).")
    print(f"The 'Value Premium' is approximately {value_premium*100:.2f}% per month.")
else:
    print(f"NO. In this period, Growth firms outperformed Value firms.")

### AUTOMATED ANSWER GENERATOR ###

QUESTION 1: Do Small firms outperform Big firms?
YES. CONFIRMED.
Data shows that Small firms (Avg: 0.0136) have higher returns than Big firms (Avg: 0.0093).
The 'Size Premium' is approximately 0.43% per month.

QUESTION 2: Do High B/M (Value) firms outperform Low B/M (Growth) firms?
YES. CONFIRMED.
Data shows that Value firms (Avg: 0.0150) have higher returns than Growth firms (Avg: 0.0089).
The 'Value Premium' is approximately 0.61% per month.


In [10]:
import pandas as pd
import numpy as np
from scipy import stats
from IPython.display import display

# ==========================================
# 1. CARICAMENTO DATI (Versione Sicura)
# ==========================================
# Portafogli
df_portfolios = pd.read_csv('25_Portfolios_5x5.csv', skiprows=15, nrows=744)
df_portfolios.rename(columns={df_portfolios.columns[0]: 'Date'}, inplace=True)
df_portfolios['Date'] = df_portfolios['Date'].astype(int)
df_portfolios.set_index('Date', inplace=True)
df_filtered = (df_portfolios.astype(float) / 100).loc[196307:202506]

# Fattori
factors = pd.read_csv('F-F_Research_Data_Factors.csv', skiprows=3, index_col=0)
factors.index = pd.to_numeric(factors.index, errors='coerce')
factors = factors[factors.index.notna()]
factors.index = factors.index.astype(int)
factors = (factors.astype(float) / 100).loc[196307:202506]

# ==========================================
# 2. CALCOLO DEI BETA
# ==========================================
aligned_data = pd.concat([factors['Mkt-RF'], df_filtered], axis=1, join='inner')
market_excess = aligned_data['Mkt-RF']
portfolios = aligned_data.drop(columns=['Mkt-RF'])
rf = factors.loc[aligned_data.index, 'RF']

betas = []
for col in portfolios.columns:
    slope, intercept, r_val, p_val, std_err = stats.linregress(market_excess, portfolios[col] - rf)
    betas.append(slope)

# ==========================================
# 3. CREAZIONE TABELLA (SENZA COLORI)
# ==========================================
table_1b_betas = pd.DataFrame(
    np.array(betas).reshape(5, 5),
    index=['Small', 'Size 2', 'Size 3', 'Size 4', 'Big'],
    columns=['Low B/M', 'B/M 2', 'B/M 3', 'B/M 4', 'High B/M']
)

print("Table 1b: Market Betas (1963:07 - 2025:06)")
# Usiamo il display semplice che non richiede librerie aggiuntive
display(table_1b_betas.round(2))

# ==========================================
# 4. CONTROLLO CAPM
# ==========================================
flat_returns = df_filtered.mean().values
flat_betas = table_1b_betas.values.flatten()
correlation = np.corrcoef(flat_returns, flat_betas)[0, 1]

print(f"\nCorrelation between Returns and Betas: {correlation:.2f}")
if correlation < 0.4:
    print("ANALYSIS: Weak correlation. High returns are NOT explained by high Beta.")
else:
    print("ANALYSIS: Positive correlation found.")

Table 1b: Market Betas (1963:07 - 2025:06)


Low B/M  B/M 2  B/M 3  B/M 4  High B/M
Small      1.46   1.27   1.18   1.10      1.13
Size 2     1.44   1.25   1.13   1.05      1.14
Size 3     1.36   1.16   1.04   0.98      1.08
Size 4     1.23   1.14   1.02   0.98      1.08
Big        0.99   0.98   0.88   0.83      0.86


Correlation between Returns and Betas: -0.08
ANALYSIS: Weak correlation. High returns are NOT explained by high Beta.
